src/hotspot/share/gc/shared/slidingForwarding.hpp

```diff
@@ -0,0 +1,179 @@
+ * last-last-ditch GC that is used when the JVM is scrambling to squeeze more space out of the heap, and at
+ * that point, ultimate performance is no longer the main concern.
+ */
+class SlidingForwarding : public CHeapObj<mtGC> {
+private:
+  static const uintptr_t MARK_LOWER_HALF_MASK = right_n_bits(32);
+
+  // We need the lowest two bits to indicate a forwarded object.
+  // The next bit indicates that the forwardee should be looked-up in a fallback-table.
+  static const int FALLBACK_SHIFT = markWord::lock_bits;
+  static const int FALLBACK_BITS = 1;
+  static const int FALLBACK_MASK = right_n_bits(FALLBACK_BITS) << FALLBACK_SHIFT;
+
+  // Next bit selects the target region
+  static const int ALT_REGION_SHIFT = FALLBACK_SHIFT + FALLBACK_BITS;
+  static const int ALT_REGION_BITS = 1;
+  // This will be "2" always, but expose it as named constant for clarity
+  static const size_t NUM_TARGET_REGIONS = 1 << ALT_REGION_BITS;
+
+  // The offset bits start then
+  static const int OFFSET_BITS_SHIFT = ALT_REGION_SHIFT + ALT_REGION_BITS;
+
+  // How many bits we use for the compressed pointer
+  static const int NUM_COMPRESSED_BITS = 32 - OFFSET_BITS_SHIFT;
+
+  // Indicates an unused base address in the target base table.
+  static HeapWord* const UNUSED_BASE;
+
+  HeapWord*  const _heap_start;
+  size_t           _num_regions;
+  size_t           _region_size_words;
+  size_t           _region_size_words_shift;
+  HeapWord**       _bases_table;
+
+  FallbackTable*   _fallback_table;
+
+  inline size_t region_index_containing(HeapWord* addr) const;
+  inline bool region_contains(HeapWord* region_base, HeapWord* addr) const;
+
+  inline uintptr_t encode_forwarding(HeapWord* from, HeapWord* to);
+  inline HeapWord* decode_forwarding(HeapWord* from, uintptr_t encoded) const;
+
+  void fallback_forward_to(HeapWord* from, HeapWord* to);
+  HeapWord* fallback_forwardee(HeapWord* from) const;
+
+public:
+  SlidingForwarding(MemRegion heap, size_t region_size_words);
+  ~SlidingForwarding();
+
+  void begin();
+  void end();
+
+  inline void forward_to(oop from, oop to);
+  inline oop forwardee(oop from) const;
+};
+
+/*
+ * A simple hash-table that acts as fallback for the sliding forwarding.
+ * This is used in the case of G1 serial compaction, which violates the
+ * assumption of sliding forwarding that each object of any region is only
+ * ever forwarded to one of two target regions. At this point, the GC is
+ * scrambling to free up more Java heap memory, and therefore performance
+ * is not the major concern.
+ *
+ * The implementation is a straightforward open hashtable.
+ * It is a single-threaded (not thread-safe) implementation, and that
+ * is sufficient because G1 serial compaction is single-threaded.
+ */
+class FallbackTable : public CHeapObj<mtGC>{
+private:
+  struct FallbackTableEntry {
+    FallbackTableEntry* _next;
+    HeapWord* _from;
+    HeapWord* _to;
+  };
+
+  static const size_t TABLE_SIZE = 128;
```

Comment:
Any reason why we do `128` here? I think we can take a bit larger table here, given that: a) the footprint would be eaten by chaining anyway; b) we delete the table after use anyway. 1K entries would take about 32K native memory, if I calculate it right.